### ISSUE TITLE
Camera distortion improvement to remove black border around distorted image

### ISSUE THREAD

**Issue by Maxx84** [osrf/gazebo#3018]
Currently, when using a camera sensor with barrel distortion, Gazebo creates an image with a black area around the distorted picture. This black area should instead contain valid pixel information.

![current_implementation_-0p1](https://user-images.githubusercontent.com/20071378/120869954-7f5e3980-c54c-11eb-949c-5412fb976a77.png)

gazebo_ros_pkgs has a _distortionCrop_ sdf tag (enabled by default) that crops the output image to remove the black border; 
however, this crop is purely cosmetic and affects the realism of camera distortion, so IMHO it should not be used and disabled by default.

I created a short document to explain how to improve the camera distortion behavior
[Gazebo_distortion_improvements.pdf](https://github.com/osrf/gazebo/files/6601567/Gazebo_distortion_improvements.pdf)

The TL;DR is that barrel distortion intrinsically increases the camera's field of view, so Gazebo's distortion plugin needs to increase the camera's field of view accordingly before acquiring a distorted image.

Here's an example image showing my proposed implementation applied to the same scene used for the image above:
![larger_fov_-0p1](https://user-images.githubusercontent.com/20071378/120870827-d6fda480-c54e-11eb-9f6a-ae8d5c07e989.png)

Unfortunately, I currently don't have time to implement the proposed changes myself, but I wanted this documented so that someone else could pick it up, if there is any interest.

**Comment by Maxx84** [comment]
Some additional notes:

1. I've designed my implementation to work with the [camera distortion fix proposed here.](https://github.com/osrf/gazebo/issues/3003) Without this fix, in step (3) the coordinates should be divided by the resolution (instead of the focal length), and then the results should be multiplied again by the resolution.
2. I think something might be off with my reverse distortion calculation, because I realized that the last image I've posted was taken with a larger FoV than the one that I calculated.
3. I realized that my proposed implementation has a flaw, the way it is implemented now. When at step (5) a new FoV is set for the camera, Gazebo automatically uses this value to internally calculate the focal length, which is then used to normalize the coordinates. I suspect that this should affect the end result of distortion. This issue would arise also without using the fix in [Issue 3003](https://github.com/osrf/gazebo/issues/3003), since also the resolution is increased.
